pkg/conversion/capi2mapi/powervs.go

```diff
@@ -0,0 +1,305 @@
+	if m.machine == nil || m.powerVSMachine == nil || m.powerVSCluster == nil {
+		return nil, nil, errCAPIMachinePowerVSMachinePowerVSClusterCannotBeNil
+	}
+
+	var (
+		warnings []string
+		errors   field.ErrorList
+	)
+	fldPath := field.NewPath("spec")
+	serviceInstance, err := convertPowerVSServiceInstanceToMAPI(fldPath.Child("serviceInstance"), m.powerVSMachine.Spec.ServiceInstanceID, m.powerVSMachine.Spec.ServiceInstance)
+	if err != nil {
+		errors = append(errors, err)
+	}
+	image, err := convertPowerVSImageToMAPI(fldPath.Child("image"), m.powerVSMachine.Spec.Image, m.powerVSMachine.Spec.ImageRef)
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	network, err := convertPowerVSNetworkToMAPI(fldPath.Child("network"), m.powerVSMachine.Spec.Network)
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	mapiProviderConfig := mapiv1.PowerVSMachineProviderConfig{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "PowerVSMachineProviderConfig",
+			APIVersion: "machine.openshift.io/v1",
+		},
+		ServiceInstance: serviceInstance,
+		Image:           image,
+		Network:         network,
+		KeyPairName:     m.powerVSMachine.Spec.SSHKey,
+		SystemType:      m.powerVSMachine.Spec.SystemType,
+		ProcessorType:   mapiv1.PowerVSProcessorType(m.powerVSMachine.Spec.ProcessorType),
+		Processors:      m.powerVSMachine.Spec.Processors,
+		MemoryGiB:       m.powerVSMachine.Spec.MemoryGiB,
+		//CredentialsSecret:
+		//LoadBalancers: TODO: Not supported for workers.
+	}
+
+	userDataSecretName := ptr.Deref(m.machine.Spec.Bootstrap.DataSecretName, "")
+	if userDataSecretName != "" {
+		mapiProviderConfig.UserDataSecret = &mapiv1.PowerVSSecretReference{
+			Name: userDataSecretName,
+		}
+	}
+
+	return &mapiProviderConfig, warnings, nil
+}
+
+// Conversion helpers.
+
+// TODO: May be we can use generics and support for all the platforms?
+// rawExtensionFromPowerVSProviderSpec marshals the machine provider spec.
+func rawExtensionFromPowerVSProviderSpec(spec *mapiv1.PowerVSMachineProviderConfig) (*runtime.RawExtension, error) {
+	if spec == nil {
+		return &runtime.RawExtension{}, nil
+	}
+
+	rawBytes, err := json.Marshal(spec)
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling providerSpec: %w", err)
+	}
+
+	return &runtime.RawExtension{
+		Raw: rawBytes,
+	}, nil
+}
+
+func convertPowerVSNetworkToMAPI(fldPath *field.Path, network capibmv1.IBMPowerVSResourceReference) (mapiv1.PowerVSResource, *field.Error) {
+	var networkResource mapiv1.PowerVSResource
+	if network.ID != nil {
+		networkResource.Type = mapiv1.PowerVSResourceTypeID
+		networkResource.ID = network.ID
+		return networkResource, nil
+	}
+	if network.Name != nil {
+		networkResource.Type = mapiv1.PowerVSResourceTypeName
+		networkResource.Name = network.Name
+		return networkResource, nil
+	}
+	if network.RegEx != nil {
+		networkResource.Type = mapiv1.PowerVSResourceTypeRegEx
+		networkResource.RegEx = network.RegEx
+		return networkResource, nil
+	}
```

Comment:
Is this ordering the same as would happen within the provider code?

Reply:
Yes, Id, Name followed by Regex.

Reply:
Added a comment.